File: 5-Problem-A-Day/84. Largest Rectangle in Histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        int ans = 0;
        for (int i = 0; i < heights.length; i++) {
            while (stack.peek() != -1 && heights[stack.peek()] >= heights[i])
                ans = Math.max(ans, heights[stack.pop()] * (i - stack.peek() - 1));
            stack.push(i);
        }
        while (stack.peek() != -1)
            ans = Math.max(ans, heights[stack.pop()] * (heights.length - stack.peek() - 1));
        return ans;
    }
}

File: 2021/84. Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int maxArea = 0;
        stack<int> hStack;
        hStack.push(-1);
        for (int i = 0; i < heights.size(); i++) {
            while (hStack.top() != -1 && heights[i] <= heights[hStack.top()]) {
                int height = heights[hStack.top()];
                hStack.pop();
                maxArea = max(maxArea, height * (i - hStack.top() - 1));
            }
            hStack.push(i);
        }
                
        
        while (hStack.top() != -1) {
            int height = heights[hStack.top()];
            hStack.pop();            
            int tmpArea = height * (heights.size() - hStack.top() - 1);
            maxArea = max(maxArea, tmpArea);
        }
        
        return maxArea;
    }
};

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> heightStack;    
        heightStack.push(-1);
        int maxArea = 0;
        for (int i = 0; i < heights.size(); i++) {
            while (heightStack.top() != -1 && heights[i] < heights[heightStack.top()]) {
                int index = heightStack.top();
                heightStack.pop();
                maxArea = max(maxArea, heights[index] * (i - heightStack.top() - 1));
            }                         
            heightStack.push(i);
        }
        while (heightStack.top() != -1) {
            int index = heightStack.top();
            heightStack.pop();
            maxArea = max(maxArea, heights[index] * ((int) heights.size() - heightStack.top() - 1));
        }
        return maxArea;
    }
};